### PR TITLE
First loop on model selection

### PR DESCRIPTION
**Pull request by aidanmattrick**
Set up processor and ran several models to determine the most effective model that we can use to perform hyperparameter optimization on. Note: I think in order to run more tests I will need to refactor some of these functions into a larger class but this should be sufficient for what we need at the moment.

### CODE REVIEW

File: src/analysis.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
+"""
+This script trains models and output results in the form of a figure to be used for further analysis.
+
+Usage: analysis.py -i=<input> -o=<output> [-v]
+
+Options:
+-i <input>, --input <input>     Local raw data csv file directory
```
Comment (zmerpez) [inline src/analysis.py]: input - filename and path
output - file directory

File: src/analysis.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+import altair as alt
+from vega_datasets import data
+
+from docopt import docopt
+from pylab import savefig
+from render_table import render_table
+
+print("start")
+args = docopt(__doc__)
+
```
Comment (zmerpez) [inline src/analysis.py:64]: Within the last scripts, we have put all prints in verbose statements. You may copy from the latest versions.

File: src/analysis.py (hunk 1 above)
Comment (aidanmattrick) [reply src/analysis.py]: Thanks Zeliha assuming you are referring to the definitions like below: 

-i <input>, --input <input>     Local processed training data csv file in directory
-o <output>, --output <output>  Local output filename and path for results table

If not let me know.

File: src/analysis.py (hunk 2 above)
Comment (aidanmattrick) [reply src/analysis.py:64]: Little confused here. I copied print statements in the main function like from the EDA file. Are you referring to this or something else?

File: src/analysis.py (hunk 1 above)
Comment (cmmclaug) [reply src/analysis.py]: In this case -i would be referring to the input processed data directory, -o would be the output directory, /results

File: src/analysis.py (hunk 2 above)
Comment (cmmclaug) [reply src/analysis.py:64]: A good reference would be the structure update PR currently in review with the standardized output procedure. The verbose flag just allows us to control whether to hide or show all prints. If you look at that review we only actually use raw 'print' statements at the beginning and end of the script to help us track full project script progress!

File: src/analysis.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+def read_input_file(input_file_path):
```
Comment (cmmclaug) [inline src/analysis.py:66]: I can modify later to update, but we may eventually want to split this into a tests validation function.

File: src/analysis.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    ----------
+        pandas.Dataframe
+
+    """
+    train_df = dataframe
+    X_train = train_df.drop(columns=['align'])
+    y_train = train_df['align']
+
+    numeric_features = ['appearances']
+    categorical_features = ['name', 'id', 'eye', 'hair', 'year', 'publisher', 'sex']
```
Comment (cmmclaug) [inline src/analysis.py]: I think we should encode year as ordinal here, as this is a sequential feature!

File: src/analysis.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    drop_features = ['gsm', 'first_appearance']
```
Comment (cmmclaug) [inline src/analysis.py:156]: Agreed on dropping gsm, it's too sparse. First appearance we can probably also leave out of our initial analysis. We would probably want to trim it to just the month to avoid duplicating year data as part of feature engineering.

File: src/analysis.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+
+    numeric_transformer = Pipeline(
+        steps=[
+            ("imputer", SimpleImputer(strategy='median')),
+            ("scaler", StandardScaler()),
+        ]
+    )
+
+    categorical_transformer = Pipeline(
```
Comment (cmmclaug) [inline src/analysis.py:165]: We may want to look at grouping the tiny groups into their own class soon.

File: src/analysis.py [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+        "Decision Tree Classifier": DecisionTreeClassifier(),
+        "Random Forest Classifier": RandomForestClassifier(),
+        "SVM": SVC(),
+        "OVR LogisticRegression": OneVsRestClassifier(LogisticRegression()),
+        "LogisticRegression": LogisticRegression(),
+    }
+
+    for key, value in models.items():
+        pipeline = Pipeline(
+                steps=[("preprocessor", preprocessor),
```
Comment (cmmclaug) [inline src/analysis.py]: Small indenting update here

File: src/analysis.py [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+    return results
+
+
+
+def save_img(data_frame, output_folder, file_name):
+    fig_1, ax_1 = render_table(data_frame, header_columns=0, col_width=5)
+    fig_1.savefig(output_folder + "/" + file_name)
+
+def main(input_file_path, output_folder_path):
+    print(input_file_path)
```
Comment (cmmclaug) [inline src/analysis.py]: Print statements and commented out code cleanup

File: src/analysis.py [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+    # print("processing data for model")
+    #train
+    model_df = train_model(data_frame)
+    print("Trained model")
+    save_img(model_df, output_folder_path, "Table_03")
+    print("Done")
+
+
+if __name__ == "__main__":
+    print("101")
```
Comment (cmmclaug) [inline src/analysis.py]: Clean up print statements here

File: src/analysis.py [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+    models = {
+        "Dummy Classifier": DummyClassifier(),
+        "Decision Tree Classifier": DecisionTreeClassifier(),
+        "Random Forest Classifier": RandomForestClassifier(),
+        "SVM": SVC(),
+        "OVR LogisticRegression": OneVsRestClassifier(LogisticRegression()),
+        "LogisticRegression": LogisticRegression(),
+    }
+
+    for key, value in models.items():
```
Comment (cmmclaug) [inline src/analysis.py:212]: this might be a good place for tqdm after!

File: src/analysis.py [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+        pipeline = Pipeline(
+                steps=[("preprocessor", preprocessor),
+               (key, value)
+                ]
+            )
+        scores = cross_validate(pipeline, X_train, y_train, return_train_score=True, n_jobs=-1, verbose=1)
```
Comment (cmmclaug) [inline src/analysis.py:219]: Do we need to run on the test set?

**Review by cmmclaug — CHANGES_REQUESTED** [review]
Mostly small updates, should be able to reference generate_eda.py to speed things up!

File: src/analysis.py (hunk 4 above)
Comment (aidanmattrick) [reply src/analysis.py]: Thanks Craig - good point

File: src/analysis.py (hunk 9 above)
Comment (aidanmattrick) [reply src/analysis.py]: Removing

File: src/analysis.py (hunk 10 above)
Comment (aidanmattrick) [reply src/analysis.py:212]: What is tqdm?

File: src/analysis.py (hunk 11 above)
Comment (aidanmattrick) [reply src/analysis.py:219]: I am running out of time relative to other labs I have to get done... As discussed per milestones thing this would be a nice to have for this week but not must have (given we have attempted hyperparameter optimization)

File: src/analysis.py (hunk 8 above)
Comment (aidanmattrick) [reply src/analysis.py]: Do you want me to remove print statements?

File: src/analysis.py (hunk 7 above)
Comment (aidanmattrick) [reply src/analysis.py]: Looking through this but not sure what is the issue. Will message on slack.

File: src/analysis.py (hunk 6 above)
Comment (aidanmattrick) [reply src/analysis.py:165]: Agreed.

File: src/analysis.py (hunk 3 above)
Comment (aidanmattrick) [reply src/analysis.py:66]: Yes, think there is room for improvement WRT to modularization.

File: src/analysis.py (hunk 2 above)
Comment (aidanmattrick) [reply src/analysis.py:64]: Thanks - put in!

File: src/analysis.py (hunk 10 above)
Comment (cmmclaug) [reply src/analysis.py:212]: It's the Python progress bar library we used in 512. It's really good and easy to use, shows what step of a script you're running and updates the progress of step graphically

File: src/analysis.py (hunk 11 above)
Comment (cmmclaug) [reply src/analysis.py:219]: Okay, if I have time later I may try to throw it in. We'll just see how it goes for now.

File: src/analysis.py (hunk 3 above)
Comment (cmmclaug) [reply src/analysis.py:66]: Haha, low priority at this point for sure.

File: src/analysis.py [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+
+Usage: analysis.py -i=<input> -o=<output> [-v]
+
+Options:
+-i <input>, --input <input>     Local processed training data csv file in directory
+-o <output>, --output <output>  Local output directory for created pngs
+[-v]                            Report verbose output of dataset retrieval process
+"""
+
+import os
```
Comment (cmmclaug) [inline src/analysis.py:16]: We probably need to review which libraries we actually need for this next milestone and trim here.

**Review by cmmclaug — APPROVED** [review]
